### DISCUSSION TITLE
Does XMake support directory scoped variables like CMake ?

### DISCUSSION THREAD

**Discussion by glenncoombs42** [xmake-io/xmake#5106] [Q&A] [answered]
I am experimenting with XMake and have the following project structure:

```
├── recorder
│   ├── win10
│   │   ├── cdr7
│   │   │   ├── project1
│   │   │   │   ├── RecorderCdr7Win10Project1.iss
│   │   │   │   └── xmake.lua
│   │   │   └── xmake.lua
│   │   ├── mdr6
│   │   │   ├── project1
│   │   │   │   ├── RecorderMdr6Win10Project1.iss
│   │   │   │   └── xmake.lua
│   │   │   └── xmake.lua
│   │   └── xmake.lua
│   ├── win7
│   │   ├── cdr7
│   │   │   ├── project1
│   │   │   │   ├── RecorderCdr7Win7Project1.iss
│   │   │   │   └── xmake.lua
│   │   │   └── xmake.lua
│   │   ├── mdr6
│   │   │   ├── project1
│   │   │   │   ├── RecorderMdr6Win7Project1.iss
│   │   │   │   └── xmake.lua
│   │   │   └── xmake.lua
│   │   └── xmake.lua
│   └── xmake.lua
└── xmake.lua
```

The top level xmake.lua looks like this:

```
rule("iscc")
    set_extensions(".iss")
    on_build_file(function (target, sourcefile, opt)

        local targetdir  = target:targetdir() .. "/"
        local targetfile = path.basename(sourcefile)

        print("IN ISCC RULE")
        print("  scriptdir     = $(scriptdir)")
        print("  targetdir     = " .. targetdir)
        print("  targetfile    = " .. targetfile)
--        print("  target_os     = " .. target_os)
    end)

target_os = "unknown"

includes("recorder/xmake.lua")
```

The xmake.lua files just include the xmake.lua files in the sub folders one level down.

Each of the xmake.lua files in the project1 folders look like this:

```
target("RecorderCdr7Win10Project1")
    on_load(function(target)
        print("In RecorderCdr7Win10Project1, on_load()")
        print("  scriptdir  = $(scriptdir)")
    end)

    set_kind("object")

    -- make the target support the construction rules of the iss script
    add_rules("iscc")

    -- add an iss script to build
    add_files("RecorderCdr7Win10Project1.iss")
```

The xmake.lua files in the win7 and win 10 folders look like this:

```
target_os = "win7"

includes("cdr7")
includes("mdr6")
```

```
target_os = "win10"

includes("cdr7")
includes("mdr6")
```

I am trying to use the `target_os` value in the `iscc` rule.  I would like this to expand to `win7` for the 2 projects in the win7 folder and to expand to `win10` for the 2 projects in the win10 folder. But when I uncomment the line to print `target_os` I get the following error:

```
error: attempt to concatenate a nil value (global 'target_os')
```

How do I access the value of `target_os` in the `iscc` rule ?  Or is there a different way of doing this ?  I am used to CMake where I can set variables in each CMakeLists.txt and they propogate to sub CMakeLists.txt files.  It seems like XMake does have a similar feature but maybe I am just not using it properly.

[xmake-test.tar.gz](https://github.com/xmake-io/xmake/files/15318955/xmake-test.tar.gz)

**Comment by waruqi** [comment]
please do not use scope variable in script scope. you can all `os.scriptdir()` instead of `${scriptdir}`

**Reply by glenncoombs42** [reply]
Thanks for the tip on `os.scriptdir()` but the problem I have is with my own variable `target_os`.  How do I access `target_os` in script scope ?

**Reply by waruqi** [reply]
lua variable? custom option?

please do not access lua variable of the description scope in script scope.

if it's custom option, you can use `get_config("xxx")`

or you can use `target:values("xxx")` and `set_values("xxx")`

```lua
local val = "xxx"
target("test")
     set_values(val)
     on_config(function (target)
         print(target:values("xxx"))
     end)

**Reply by glenncoombs42** [reply]
`target_os` is a global lua variable declared in the top level xmake.lua.  I set it to `win7` in the `recorder/win7/xmake.lua` file, and I set it to `win10` in the `recorder/win10/xmake.lua` file. I try to use it in the `iscc` rule declared in the top level xmake.lua file but it fails with the error message `error: attempt to concatenate a nil value (global 'target_os')`.

For the targets defined in xmake.lua files under the `recorder/win7` folder I want the `iscc` rule to pass `win7` to the iscc tool.  And for the targets defined in xmake.lua files under the `recorder/win10` folder I want the `iscc` rule to pass `win10` to the iscc tool.

Maybe I should not be using a global variable.  How would you set a value in one xmake.lua file that is inherited by all xmake.lua files that are subsequently included using `includes` ?  So the file where the `target_os` value is set is different to the files where the targets that need to use `target_os` are defined.

**Comment by waruqi** [comment, marked as answer]
> lua variable? custom option?
> 
> please do not access lua variable of the description scope in script scope.
> 
> if it's custom option, you can use `get_config("xxx")`
> 
> or you can use `target:values("xxx")` and `set_values("xxx")`
> 
> ```lua
> local val = "xxx"
> target("test")
>      set_values(val)
>      on_config(function (target)
>          print(target:values("xxx"))
>      end)
> ```

try set_values

**Reply by glenncoombs42** [reply]
Thank you, using `set_values` does work.  I have modified `recorder/win10/cdr7/project1/xmake.lua` to look like this:

```
-- the next line must be uncommented if target_os is needed inside the on_load() function
-- local target_os = target_os

target("RecorderCdr7Win10Project1")
    on_load(function(target)
        print("In RecorderCdr7Win10Project1, on_load()")
        print("  scriptdir  = ", os.scriptdir())
        print("  target_os = ", target_os)
    end)

    set_kind("object")

    -- make the target support the construction rules of the iss script
    add_rules("iscc")

    set_values("iscc.flags", target_os)

    -- add an iss script to build
    add_files("RecorderCdr7Win10Project1.iss")
```

I notice that I cannot reference `target_os` inside the `on_load()` function unless I copy `target_os` to a local copy at the start of the file.  Is this a deliberate restriction ?  Why can I see `target_os` in `description scope` but not in `script scope` ?
